datatransfer/impl/graphsync/graphsync_impl.go

```diff
@@ -57,15 +61,74 @@ func NewGraphSyncDataTransfer(parent context.Context, host host.Host, gs graphsy
 		nil,
 		make(map[string]validateType),
 		make(map[datatransfer.ChannelID]datatransfer.ChannelState),
+		sync.Mutex{},
 		gs,
 		host.ID(),
 		0,
 	}
+	if err := gs.RegisterRequestReceivedHook(true, impl.GsReqRecdHook); err != nil {
+		log.Error(err)
+		return nil
+	}
 	receiver := &graphsyncReceiver{parent, impl}
 	dataTransferNetwork.SetDelegate(receiver)
 	return impl
 }
 
+// GsReqRecdHook is a graphsync.OnRequestReceivedHook hook
```

Comment:
this should not be a public function.
if it needs to be tested individually, it should be extracted to another module